File: Util/Util2/TaskUtil.fs
module TaskUtil
open System.Threading.Tasks;
open System.Threading;

type TaskHelper =
    static member Create<'T>(mainLogic,(action:System.Action<Task<'T>>))  =
        Task.Factory.StartNew(mainLogic).ContinueWith(action, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.FromCurrentSynchronizationContext())

    static member Create((mainLogic:System.Action),(action:System.Action<Task>)) =
        Task.Factory.StartNew(mainLogic).ContinueWith(action,CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.FromCurrentSynchronizationContext())    

    static member Create((mainLogic:System.Action<obj>),(action:System.Action<Task>),(state:obj))  =
        Task.Factory.StartNew(mainLogic,state).ContinueWith(action,CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.FromCurrentSynchronizationContext())

    static member Create<'T>(mainLogic,(action:System.Action<Task<'T>>),(state:obj)) =
        Task.Factory.StartNew<'T>(mainLogic,state).ContinueWith(action,CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.FromCurrentSynchronizationContext())

File: Util/Util/TaskUtil.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;

namespace Util
{
   public static class TaskUtil
    {
        public static Task TaskHelper<T>(Func<T> mainLogin, Action<Task<T>> continueAction)
        {
            Task task = Task.Factory.StartNew<T>(mainLogin)
                      .ContinueWith(continueAction, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.FromCurrentSynchronizationContext());
            return task;
        }

        public static Task TaskHelper(Action mainLogin, Action<Task> continueAction)
        {
            Task task = Task.Factory.StartNew(mainLogin)
                      .ContinueWith(continueAction, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.FromCurrentSynchronizationContext());
            return task;
        }
        public static Task TaskHelper(Action<object> mainLogin, Action<Task> continueAction, object state)
        {
            Task task = Task.Factory.StartNew(mainLogin, state)
                      .ContinueWith(continueAction, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.FromCurrentSynchronizationContext());
            return task;

        }

        public static Task TaskHelper<T>(Func<object, T> mainLogin, Action<Task<T>> continueAction, object state)
        {
            Task task = Task.Factory.StartNew<T>(mainLogin, state)
                      .ContinueWith(continueAction, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.FromCurrentSynchronizationContext());
            return task;
        }
    }
}
